### PR TITLE
caddyhttp: Support configuring Server from handler provisioning

### PR DESCRIPTION
**Pull request by WeidiDeng**
Fixes [4765](https://github.com/caddyserver/caddy/issues/4765).

When Middleware handler is provisioning, it can access caddyhttp.Server and register specific http.Server related callbacks.

### CODE REVIEW

**Comment by mholt** [comment]
Thanks, I'll look at this soon!

File: modules/caddyhttp/server.go [hunk 1]
```diff
@@ -146,6 +146,56 @@ type Server struct {
...
+	s.connContexts = append(s.connContexts, f)
+}
+
+// register OnShutdown callbacks
+func (s *Server) RegisterOnShutdown(f func()) {
+	s.onShutdowns = append(s.onShutdowns, f)
+}
+
+// apply callbacks to server
+func (s *Server) ConfigureServer(server *http.Server) {
```
Comment (mholt) [inline modules/caddyhttp/server.go:172]: Does this have to be exported? It should only be configured within this package, right?

File: modules/caddyhttp/app.go [hunk 2]
```diff
@@ -173,7 +173,9 @@ func (app *App) Provision(ctx caddy.Context) error {
 	}
 
 	// prepare each server
+	oldContext := ctx.Context
 	for srvName, srv := range app.Servers {
+		ctx.Context = context.WithValue(oldContext, ServerCtxKey, srv)
```
Comment (mholt) [inline modules/caddyhttp/app.go:178]: What is the purpose of this?

**Review by mholt — CHANGES_REQUESTED** [review]
Thanks for the enhancement! This looks mostly good, though I do have a couple questions.

File: modules/caddyhttp/server.go (hunk 1 above)
Comment (WeidiDeng) [reply modules/caddyhttp/server.go:172]: Yes, it should be unexported.

File: modules/caddyhttp/app.go (hunk 2 above)
Comment (WeidiDeng) [reply modules/caddyhttp/app.go:178]: In http middleware provision, this gives them to access current server by using context key. Using context value to get current server the middleware is handling request.

**Review by mholt — APPROVED** [review]
I went ahead and unexported that method, and just moved a couple things around; this looks useful, so I will be happy to merge it in. Thanks @WeidiDeng!

**Pull request merged**